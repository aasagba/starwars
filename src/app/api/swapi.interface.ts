export interface IPeopleResponse {
  count: number,
  next: string,
  previous: string,
  results: Array<IPeople>
}

export interface IPeople {
  birth_year: string,
  eye_color: string,
  films: Array<string>,
  gender: string,
  hair_color: string,
  height: string,
  homeworld: string,
  mass: string,
  name: string,
  skin_color: string,
  created: Date,
  edited: Date,
  species: Array<string>,
  starships: Array<string>,
  url: string,
  vehicles: Array<string>
}

export interface ISearchParams {
  searchTerm?: string
}

