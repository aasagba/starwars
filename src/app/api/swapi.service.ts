import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';

import { environment } from '../../environments/environment';
import { Observable, of } from 'rxjs';
import { IPeopleResponse, ISearchParams } from './swapi.interface';
import { catchError, map } from 'rxjs/operators';

@Injectable({
  providedIn: 'root'
})
export class SwapiService {

  constructor(private httpClient: HttpClient){}

  public getStarWarsPeople(searchParams?: ISearchParams):
    Observable<IPeopleResponse | Error> {

    let params: HttpParams = new HttpParams();

    if (searchParams && searchParams.searchTerm) {
      params = params.append('search', searchParams.searchTerm);
    }

    if (searchParams && searchParams.page) {
      params = params.append('page', searchParams.page);
    }

    return this.httpClient
      .get(`${environment.baseUrl}/people`, { params })
      .pipe(map((response: IPeopleResponse) => response),
        catchError((error: Error) => of(new Error(error.message)))
      );
  }

}
