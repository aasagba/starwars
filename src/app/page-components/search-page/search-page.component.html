<form [formGroup]="searchPageForm"
      class="clr-form">
  <app-search (searchEvent)="search($event)" [searchTerm]="searchTerm"></app-search>
</form>

<clr-datagrid [clrDgLoading]="(isLoading$ | async)"
              (clrDgRefresh)="refresh($event)" >

  <clr-dg-placeholder
    [hidden]="(people$ | async)?.length !== 0">
    <span>No Items to display</span>
  </clr-dg-placeholder>

  <clr-dg-column>Name</clr-dg-column>
  <clr-dg-column>Height</clr-dg-column>
  <clr-dg-column>Mass</clr-dg-column>
  <clr-dg-column>Hair Colour</clr-dg-column>
  <clr-dg-column>Eye Colour</clr-dg-column>
  <clr-dg-column>Birth Year</clr-dg-column>
  <clr-dg-column>Gender</clr-dg-column>
  <clr-dg-column>Homeworld</clr-dg-column>
  <clr-dg-column>Films</clr-dg-column>
  <clr-dg-column>Species</clr-dg-column>
  <clr-dg-column>Vehicles</clr-dg-column>
  <clr-dg-column>Starships</clr-dg-column>

  <clr-dg-row *ngFor="let person of (people$ | async); index as i;">
    <clr-dg-cell>{{ person.name }}</clr-dg-cell>
    <clr-dg-cell>{{ person.height }}</clr-dg-cell>
    <clr-dg-cell>{{ person.mass }}</clr-dg-cell>
    <clr-dg-cell>{{ person.hair_color }}</clr-dg-cell>
    <clr-dg-cell>{{ person.eye_color }}</clr-dg-cell>
    <clr-dg-cell>{{ person.birth_year }}</clr-dg-cell>
    <clr-dg-cell>{{ person.gender }}</clr-dg-cell>
    <clr-dg-cell>{{ person.homeworld }}</clr-dg-cell>
    <clr-dg-cell>{{ person.films?.length }}</clr-dg-cell>
    <clr-dg-cell>{{ person.species?.length }}</clr-dg-cell>
    <clr-dg-cell>{{ person.vehicles?.length }}</clr-dg-cell>
    <clr-dg-cell>{{ person.starships?.length }}</clr-dg-cell>
  </clr-dg-row>

  <clr-dg-footer>
    <clr-dg-pagination #pagination [clrDgPageSize]="10" [clrDgTotalItems]="total$ | async">
      <clr-dg-page-size [clrPageSizeOptions]="[10]">Per page</clr-dg-page-size>
      {{pagination.firstItem + 1}} - {{pagination.lastItem + 1}}
      of {{pagination.totalItems}} People
    </clr-dg-pagination>
  </clr-dg-footer>

</clr-datagrid>

