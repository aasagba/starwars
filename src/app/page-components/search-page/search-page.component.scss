:host {

  .search-wrapper {
    display: flex;
    align-items: center;
    justify-content: center;
    flex-wrap: wrap;
    flex-direction: column;
  }

  .clr-form {
    .search {
      margin-bottom: 34px;

      ::ng-deep .clr-control-container {
        align-self: center;

        input {
          width: 8rem;
          color: black;
          text-align: center;
        }
      }
    }
  }

}
