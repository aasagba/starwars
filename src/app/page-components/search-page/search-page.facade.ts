import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Store } from '@ngrx/store';
import { State } from '../../reducers';
import { SearchPeopleAction } from './ngrx/actions';
import { IPeople } from '../../api/swapi.interface';
import { SearchSelectors } from './ngrx/selectors';

@Injectable()
export class SearchPageFacade {
  public isLoading$: Observable<boolean>;
  public people$: Observable<Array<IPeople>>;

  constructor(private store: Store<State>) {
    this.isLoading$ = this.store.select(SearchSelectors.isLoading);
    this.people$ = this.store.select(SearchSelectors.people);
  }

  public searchPeople(): void {
    this.store.dispatch(new SearchPeopleAction())
  }
}
