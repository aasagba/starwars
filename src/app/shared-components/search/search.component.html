<div class="clr-row">
  <div class="clr-col-12">
    <div class="search-wrapper"
         [formGroup]="searchGroup">

      <div class="logo"></div>

      <clr-input-container class="search">
        <input
          clrInput
          class="clr-input"
          type="search"
          id="search"
          name="search"
          formControlName="search"
          placeholder="Search People"
          (search)="search()" />
      </clr-input-container>
    </div>
  </div>
</div>

