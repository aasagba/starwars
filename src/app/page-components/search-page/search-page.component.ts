import { AfterViewInit, Component, OnInit, ViewChild } from '@angular/core';
import { FormControl, FormGroup } from '@angular/forms';
import { Observable } from 'rxjs';
import { IPeople, ISearchParams } from '../../api/swapi.interface';
import { SearchPageFacade } from './search-page.facade';
import { ClrDatagridPagination, ClrDatagridStateInterface } from '@clr/angular';
import { ActivatedRoute, ParamMap, Router } from '@angular/router';

@Component({
  selector: 'app-search-page',
  templateUrl: './search-page.component.html',
  styleUrls: ['./search-page.component.scss']
})
export class SearchPageComponent implements OnInit, AfterViewInit {
  public searchPageForm: FormGroup;
  public isLoading$: Observable<boolean>;
  public people$: Observable<Array<IPeople>>;
  public total$: Observable<number>;
  public searchTerm = '';

  @ViewChild('pagination')
  pagination: ClrDatagridPagination;

  constructor(private facade: SearchPageFacade, private route: ActivatedRoute, private router: Router) {
    // this.route.params.subscribe( params => console.log(params) );
  }

  ngOnInit() {
    this.isLoading$ = this.facade.isLoading$;
    this.people$ = this.facade.people$;
    this.total$ = this.facade.count$;
    this.setupForm();

    this.route.paramMap.subscribe(
      (params: ParamMap) => {
        this.searchTerm = params.get('term');
        console.log(`ngAfterViewInit searchTerm ${this.searchTerm}`);
        this.facade.searchPeople(
          { searchTerm: this.searchTerm });
      }
    );
  }

  public search(searchTerm: string): void {
    // this.facade.searchPeople({ searchTerm });
    this.router.navigateByUrl(`/search/${searchTerm}`);
    // reset pagination currentpage
    this.pagination.currentPage = 1;
  }

  public refresh(state: ClrDatagridStateInterface): void {
    console.log('state', state);
    console.log('refresh');
    const searchTerm: string =
      this.searchPageForm.controls.search.value;

    const searchParams: ISearchParams = {
      searchTerm,
      page: state.page
        ? state.page.current.toString()
        : ''
    };

    this.facade.searchPeople(searchParams);
  }

  private setupForm(): void {
    this.searchPageForm = new FormGroup({
      search: new FormControl('')
    });
  }

  ngAfterViewInit(): void {
  }

}
