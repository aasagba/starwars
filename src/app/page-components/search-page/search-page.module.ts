import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ClarityModule } from '@clr/angular';
import { SearchPageComponent } from './search-page.component';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { SearchPageFacade } from './search-page.facade';
import { SearchPeopleEffect } from './ngrx/effects';
import { EffectsModule } from '@ngrx/effects';
import { SearchModule } from '../../shared-components/search/search.module';

@NgModule({
  imports: [
    EffectsModule.forFeature([
      SearchPeopleEffect
    ]),
    CommonModule,
    ClarityModule,
    ReactiveFormsModule,
    FormsModule,
    SearchModule
  ],
  declarations: [
    SearchPageComponent
  ],
  providers: [SearchPageFacade],
  exports: [SearchPageComponent]
})
export class SearchPageModule {}
