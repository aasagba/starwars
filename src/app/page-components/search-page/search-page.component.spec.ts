import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { SearchPageComponent } from './search-page.component';
import { SearchPageFacade } from './search-page.facade';
import { BehaviorSubject } from 'rxjs';
import { IPeople, ISearchParams } from '../../api/swapi.interface';
import { mockPeople } from '../../../../test/mock/api/swapi-mock';
import { ClarityModule, ClrDatagrid, ClrDatagridStateInterface } from '@clr/angular';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { By } from '@angular/platform-browser';
import { SearchModule } from '../../shared-components/search/search.module';

describe('SearchPageComponent', () => {
  const isLoading$: BehaviorSubject<boolean> = new BehaviorSubject(false);
  const people$: BehaviorSubject<Array<IPeople>> = new BehaviorSubject(mockPeople);
  const count$: BehaviorSubject<number> = new BehaviorSubject(0);

  let component: SearchPageComponent;
  let fixture: ComponentFixture<SearchPageComponent>;
  let facade: SearchPageFacade;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [ SearchPageComponent ],
      imports: [
        ClarityModule,
        ReactiveFormsModule,
        SearchModule,
        FormsModule
      ],
      providers: [
        {
          provide: SearchPageFacade,
          useValue: {
            searchPeople: jest.fn(),
            isLoading$: isLoading$.asObservable(),
            people$: people$.asObservable(),
            count$: count$.asObservable()
          }
        }
      ]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(SearchPageComponent);
    component = fixture.componentInstance;
    facade = TestBed.get(SearchPageFacade);
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('datagrid', () => {
    it('should be defined', () => {
      const dataGridCmp: ClrDatagrid = fixture.debugElement
        .query(By.directive(ClrDatagrid))
        .componentInstance;
      expect(dataGridCmp).toBeDefined();
    });

    it('should show placeholder and hide column titles is there is no data', () => {
      people$.next([]);
      fixture.detectChanges();
      const placeholder: ClrDatagrid = fixture.debugElement
        .query(By.directive(ClrDatagrid)).componentInstance;

      expect(placeholder).toBeTruthy();
    });
  });

  describe('search', () => {
    beforeEach(() => {
      people$.next(mockPeople);
    });

    it('it should call search on facade', () => {
      jest.spyOn(facade, 'searchPeople');

      component.search('Han Solo');
      expect(facade.searchPeople).toHaveBeenCalledWith({ searchTerm: 'Han Solo' });
    });

    it('should update datagrid with search results', () => {
      component.search('luke');

      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should reset pagination currentPage', () => {
      component.search('luke');
      expect(component.pagination.currentPage === 1);
    });
  });

  describe('pagination', () => {
    it('it should call searchPeople on the facade on a datagrid refresh event', () => {
      jest.spyOn(facade, 'searchPeople');

      const refreshState: ClrDatagridStateInterface = {
        page: {
          current: 1
        }
      };

      const expectedSearchParams: ISearchParams = {
        searchTerm: 'r2-d2',
        page: '1'
      };

      component.searchPageForm.controls.search.setValue('r2-d2');
      component.refresh(refreshState);
      expect(facade.searchPeople).toHaveBeenCalledWith(expectedSearchParams);
    });
  });


});
