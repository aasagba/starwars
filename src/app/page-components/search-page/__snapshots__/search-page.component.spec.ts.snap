// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPageComponent search should update datagrid with search results 1`] = `
<div
  id="root4"
>
  <form
    class="clr-form ng-untouched ng-pristine ng-valid"
    novalidate=""
  >
    <app-search>
      <div
        class="clr-row"
      >
        <div
          class="clr-col-12"
        >
          <div
            class="search-wrapper ng-untouched ng-pristine ng-valid"
          >
            <div
              class="logo"
            />
            <clr-input-container
              class="search clr-form-control"
            >
              
              <div
                class="clr-control-container clr-input"
              >
                <div
                  class="clr-input-wrapper"
                >
                  <input
                    aria-describedby="search-helper"
                    class="clr-input ng-untouched ng-pristine ng-valid"
                    clrinput=""
                    formcontrolname="search"
                    id="search"
                    name="search"
                    placeholder="Search People"
                    type="search"
                  />
                  
                </div>
                
                
                
              </div>
            </clr-input-container>
          </div>
        </div>
      </div>
    </app-search>
  </form>
  <clr-datagrid
    class="datagrid-host"
  >
    <div
      class="datagrid"
    >
      <div
        class="datagrid-table-wrapper"
      >
        <div
          class="datagrid-table"
          role="grid"
          tabindex="-1"
        >
          <div
            class="datagrid-header"
            role="rowgroup"
          >
            <div
              class="datagrid-row"
              role="row"
            >
              <div
                class="datagrid-row-master datagrid-row-flex"
              >
                <div
                  class="datagrid-row-sticky"
                />
                <div
                  class="datagrid-row-scrollable"
                >
                  
                  
                  
                  
                  
                  <clr-dg-column
                    aria-sort="none"
                    class="datagrid-column"
                    role="columnheader"
                    style="width: 0px;"
                  >
                    <div
                      class="datagrid-column-flex"
                    >
                      
                      
                      
                      
                      <span
                        class="datagrid-column-title"
                      >
                        
                        Name
                        
                      </span>
                      <clr-dg-column-separator
                        class="datagrid-column-separator"
                      >
                        <div
                          aria-hidden="true"
                          class="datagrid-column-handle drag-handle draggable"
                          clrdraggable=""
                        />
                        <div
                          class="datagrid-column-resize-tracker"
                        />
                      </clr-dg-column-separator>
                    </div>
                  </clr-dg-column>
                  
                  <clr-dg-column
                    aria-sort="none"
                    class="datagrid-column"
                    role="columnheader"
                    style="width: 0px;"
                  >
                    <div
                      class="datagrid-column-flex"
                    >
                      
                      
                      
                      
                      <span
                        class="datagrid-column-title"
                      >
                        
                        Height
                        
                      </span>
                      <clr-dg-column-separator
                        class="datagrid-column-separator"
                      >
                        <div
                          aria-hidden="true"
                          class="datagrid-column-handle drag-handle draggable"
                          clrdraggable=""
                        />
                        <div
                          class="datagrid-column-resize-tracker"
                        />
                      </clr-dg-column-separator>
                    </div>
                  </clr-dg-column>
                  
                  <clr-dg-column
                    aria-sort="none"
                    class="datagrid-column"
                    role="columnheader"
                    style="width: 0px;"
                  >
                    <div
                      class="datagrid-column-flex"
                    >
                      
                      
                      
                      
                      <span
                        class="datagrid-column-title"
                      >
                        
                        Mass
                        
                      </span>
                      <clr-dg-column-separator
                        class="datagrid-column-separator"
                      >
                        <div
                          aria-hidden="true"
                          class="datagrid-column-handle drag-handle draggable"
                          clrdraggable=""
                        />
                        <div
                          class="datagrid-column-resize-tracker"
                        />
                      </clr-dg-column-separator>
                    </div>
                  </clr-dg-column>
                  
                  <clr-dg-column
                    aria-sort="none"
                    class="datagrid-column"
                    role="columnheader"
                    style="width: 0px;"
                  >
                    <div
                      class="datagrid-column-flex"
                    >
                      
                      
                      
                      
                      <span
                        class="datagrid-column-title"
                      >
                        
                        Hair Colour
                        
                      </span>
                      <clr-dg-column-separator
                        class="datagrid-column-separator"
                      >
                        <div
                          aria-hidden="true"
                          class="datagrid-column-handle drag-handle draggable"
                          clrdraggable=""
                        />
                        <div
                          class="datagrid-column-resize-tracker"
                        />
                      </clr-dg-column-separator>
                    </div>
                  </clr-dg-column>
                  
                  <clr-dg-column
                    aria-sort="none"
                    class="datagrid-column"
                    role="columnheader"
                    style="width: 0px;"
                  >
                    <div
                      class="datagrid-column-flex"
                    >
                      
                      
                      
                      
                      <span
                        class="datagrid-column-title"
                      >
                        
                        Eye Colour
                        
                      </span>
                      <clr-dg-column-separator
                        class="datagrid-column-separator"
                      >
                        <div
                          aria-hidden="true"
                          class="datagrid-column-handle drag-handle draggable"
                          clrdraggable=""
                        />
                        <div
                          class="datagrid-column-resize-tracker"
                        />
                      </clr-dg-column-separator>
                    </div>
                  </clr-dg-column>
                  
                  <clr-dg-column
                    aria-sort="none"
                    class="datagrid-column"
                    role="columnheader"
                    style="width: 0px;"
                  >
                    <div
                      class="datagrid-column-flex"
                    >
                      
                      
                      
                      
                      <span
                        class="datagrid-column-title"
                      >
                        
                        Birth Year
                        
                      </span>
                      <clr-dg-column-separator
                        class="datagrid-column-separator"
                      >
                        <div
                          aria-hidden="true"
                          class="datagrid-column-handle drag-handle draggable"
                          clrdraggable=""
                        />
                        <div
                          class="datagrid-column-resize-tracker"
                        />
                      </clr-dg-column-separator>
                    </div>
                  </clr-dg-column>
                  
                  <clr-dg-column
                    aria-sort="none"
                    class="datagrid-column"
                    role="columnheader"
                    style="width: 0px;"
                  >
                    <div
                      class="datagrid-column-flex"
                    >
                      
                      
                      
                      
                      <span
                        class="datagrid-column-title"
                      >
                        
                        Gender
                        
                      </span>
                      <clr-dg-column-separator
                        class="datagrid-column-separator"
                      >
                        <div
                          aria-hidden="true"
                          class="datagrid-column-handle drag-handle draggable"
                          clrdraggable=""
                        />
                        <div
                          class="datagrid-column-resize-tracker"
                        />
                      </clr-dg-column-separator>
                    </div>
                  </clr-dg-column>
                  
                  <clr-dg-column
                    aria-sort="none"
                    class="datagrid-column"
                    role="columnheader"
                    style="width: 0px;"
                  >
                    <div
                      class="datagrid-column-flex"
                    >
                      
                      
                      
                      
                      <span
                        class="datagrid-column-title"
                      >
                        
                        Homeworld
                        
                      </span>
                      <clr-dg-column-separator
                        class="datagrid-column-separator"
                      >
                        <div
                          aria-hidden="true"
                          class="datagrid-column-handle drag-handle draggable"
                          clrdraggable=""
                        />
                        <div
                          class="datagrid-column-resize-tracker"
                        />
                      </clr-dg-column-separator>
                    </div>
                  </clr-dg-column>
                  
                  <clr-dg-column
                    aria-sort="none"
                    class="datagrid-column"
                    role="columnheader"
                    style="width: 0px;"
                  >
                    <div
                      class="datagrid-column-flex"
                    >
                      
                      
                      
                      
                      <span
                        class="datagrid-column-title"
                      >
                        
                        Films
                        
                      </span>
                      <clr-dg-column-separator
                        class="datagrid-column-separator"
                      >
                        <div
                          aria-hidden="true"
                          class="datagrid-column-handle drag-handle draggable"
                          clrdraggable=""
                        />
                        <div
                          class="datagrid-column-resize-tracker"
                        />
                      </clr-dg-column-separator>
                    </div>
                  </clr-dg-column>
                  
                  <clr-dg-column
                    aria-sort="none"
                    class="datagrid-column"
                    role="columnheader"
                    style="width: 0px;"
                  >
                    <div
                      class="datagrid-column-flex"
                    >
                      
                      
                      
                      
                      <span
                        class="datagrid-column-title"
                      >
                        
                        Species
                        
                      </span>
                      <clr-dg-column-separator
                        class="datagrid-column-separator"
                      >
                        <div
                          aria-hidden="true"
                          class="datagrid-column-handle drag-handle draggable"
                          clrdraggable=""
                        />
                        <div
                          class="datagrid-column-resize-tracker"
                        />
                      </clr-dg-column-separator>
                    </div>
                  </clr-dg-column>
                  
                  <clr-dg-column
                    aria-sort="none"
                    class="datagrid-column"
                    role="columnheader"
                    style="width: 0px;"
                  >
                    <div
                      class="datagrid-column-flex"
                    >
                      
                      
                      
                      
                      <span
                        class="datagrid-column-title"
                      >
                        
                        Vehicles
                        
                      </span>
                      <clr-dg-column-separator
                        class="datagrid-column-separator"
                      >
                        <div
                          aria-hidden="true"
                          class="datagrid-column-handle drag-handle draggable"
                          clrdraggable=""
                        />
                        <div
                          class="datagrid-column-resize-tracker"
                        />
                      </clr-dg-column-separator>
                    </div>
                  </clr-dg-column>
                  
                  <clr-dg-column
                    aria-sort="none"
                    class="datagrid-column"
                    role="columnheader"
                    style="width: 0px;"
                  >
                    <div
                      class="datagrid-column-flex"
                    >
                      
                      
                      
                      
                      <span
                        class="datagrid-column-title"
                      >
                        
                        Starships
                        
                      </span>
                      <clr-dg-column-separator
                        class="datagrid-column-separator"
                      >
                        <div
                          aria-hidden="true"
                          class="datagrid-column-handle drag-handle draggable"
                          clrdraggable=""
                        />
                        <div
                          class="datagrid-column-resize-tracker"
                        />
                      </clr-dg-column-separator>
                    </div>
                  </clr-dg-column>
                  
                </div>
              </div>
            </div>
          </div>
          
          <clr-dg-row
            aria-owns="clr-dg-row7"
            class="datagrid-row"
            role="rowgroup"
          >
            
            
            
            <div
              class="datagrid-row-master datagrid-row-flex"
              id="clr-dg-row7"
              role="row"
            >
              <div
                class="datagrid-row-sticky"
              >
                
              </div>
              <div
                class="datagrid-row-scrollable"
              >
                <div
                  class="datagrid-scrolling-cells"
                >
                  
                  
                  
                  
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    Luke Skywalker
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    172
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    blond
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                </div>
                
              </div>
              
              
              
              
            </div>
            
            
            
            
          </clr-dg-row>
          
          <clr-dg-row
            aria-owns="clr-dg-row8"
            class="datagrid-row"
            role="rowgroup"
          >
            
            
            
            <div
              class="datagrid-row-master datagrid-row-flex"
              id="clr-dg-row8"
              role="row"
            >
              <div
                class="datagrid-row-sticky"
              >
                
              </div>
              <div
                class="datagrid-row-scrollable"
              >
                <div
                  class="datagrid-scrolling-cells"
                >
                  
                  
                  
                  
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    Han Solo
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    182
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    brown
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                  <clr-dg-cell
                    class="datagrid-cell"
                    role="gridcell"
                    style="width: 0px;"
                  >
                    
                  </clr-dg-cell>
                  
                </div>
                
              </div>
              
              
              
              
            </div>
            
            
            
            
          </clr-dg-row>
          
          <clr-dg-placeholder
            class="datagrid-placeholder-container"
            hidden=""
          >
            <div
              class="datagrid-placeholder"
            >
              
              
            </div>
          </clr-dg-placeholder>
          
        </div>
      </div>
    </div>
    <clr-dg-footer
      class="datagrid-footer"
    >
      
      
      <div
        class="datagrid-footer-description"
      />
      <clr-dg-pagination
        class="pagination"
      >
        
        <div
          class="pagination-size"
        >
          <clr-dg-page-size>
            Per page
            <div
              class="clr-select-wrapper"
            >
              <select
                class="clr-page-size-select ng-untouched ng-pristine ng-valid"
              >
                
                <option
                  value="0: 10"
                >
                  10
                </option>
              </select>
            </div>
          </clr-dg-page-size>
        </div>
        <div
          class="pagination-description"
        >
           1 - 10 of 0 People 
        </div>
        
      </clr-dg-pagination>
    </clr-dg-footer>
    
    <div
      class="datagrid-calculation-table"
    >
      <div
        class="datagrid-calculation-header"
      >
        
      </div>
      
    </div>
  </clr-datagrid>
</div>
`;
